Uniforms in asm and some more

digetx opened the pull request:
One step closer to a cube-textured in assembly 🚀

digetx committed:
libhost1x: add host1x_bo_flush() and use where necessary

host1x_bo_invalidate() marks mmap'ed range as "read uncached", that's not
what we want when we are writing to memory.

digetx committed:
libhost1x: remove nvhost-display.h nvhost-gr2d.h nvhost-gr3d.h

Move theirs stuff to nvhost.h, it's quite small.

digetx committed:
libhost1x: add nvhost_framebuffer_init()

BLOB isnt't zero'ing allocated memory, resulting in a garbage being displayed
for a moment on start. Zero it in grate.

digetx committed:
grate: vertex asm: support uniforms

Vertex shader uniforms, defined in the assembly, are exposed to the shader user.

Example:

.uniforms
	[33] = "uniform_33";

digetx committed:
tools/assembler: parse vertex uniforms cmd line arguments

And set vertex uniforms in the shader program.

digetx committed:
tests/grate: asm: add vertex uniform test

digetx committed:
host1x.h: cleanup whitespaces

digetx committed:
grate: support fragment uniforms

Fragment shader uniforms, defined in the assembly, are exposed to the shader user.

Example:

.uniforms
	[1]   = "uniform_1_fp20";
	[2].l = "uniform_2_fx10_low";
	[2].h = "uniform_2_fx10_high";

The BLOB's symbol location encoding is reused, so grate_3d_ctx_set_fragment_uniform()
also works for the cgc-generated shaders.

digetx committed:
grate: more grate_error()

Replace remaining fprint's with grate_error() for consistency.

digetx committed:
tools/assembler: parse fragment uniforms cmd line arguments

And set fragment uniforms in the shader program.

digetx committed:
grate: fix count of fragment uniforms in a shader program

There was a copy-paste typo in the grate_program_add_uniform().

digetx committed:
tests/grate: asm: add fragment uniform test

digetx committed:
pass BO offset in HOST1X_BO_*()

For consistency.

digetx committed:
grate: add grate_shader_parse_*_asm_from_file()

Add asm shader load-and-parse helper routines.

digetx committed:
tools/assembler: use grate_shader_parse_*_asm_from_file() helpers

kusma commented on src/libgrate/grate-asm.c:
```diff
@@ -398,6 +398,46 @@ struct grate_shader *grate_shader_parse_fragment_asm(const char *asm_txt)
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FX10_HIGH;
+			break;
+		case FS_UNIFORM_FP20:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FP20;
```
This seems like you're setting up a symobol per scalar value. Is that so? If so, I don't think it's a great idea.

I think it'd be better if we used the types already defined in glsl_type, and augmented it with a precision-value. We could use location to device between GRATE_TYPE_FX10_LOW and GRATE_TYPE_FX10_HIGH, by requiring GRATE_TYPE_FP20 to be aligned to multiples of two, and letting the last bit of location as a differentiator for FX10...

kusma replied:
By the way, it'd be totally fine to just support GLSL_TYPE_FLOAT as a first approach.

kusma replied:
And we might not want to keep the GLSL-prefixes here. There's no point in differentiating between GLSL-types and other types. IMO...

kusma replied:
The reason why I care about this, is that I don't want combinatory explosion here. Types like this explode very quickly...

In fact, we might want to have something like vector/matrix-sizes... Hmm, dunno, maybe I'm over-thinking this now...

digetx replied:
Thank you very much for the suggestions. BLOB seems encode fragment uniform type-position in the symbol location, maybe we can re-use it. However, the type enum seem differ from the vertex uniforms.

kusma commented:
Apart from the detail above, I think this is pretty nice :)

digetx committed:
tools/assembler: move grate_framebuffer_data()

Otherwise a staled data may be retrieved from the CPU cache.

digetx committed:
tools/assembler: cleanup getopt_long format

Help option doesn't require an argument.

digetx commented:
I corrected "libhost1x: add host1x_bo_flush()" commit a bit and and added couple minor updates for  	tools/assembler. Will do something about fragment uniforms ASAP 🤒

digetx commented:
I squashed "grate: refactor grate_3d_ctx_set_fragment_uniform*()" and "grate: fragment asm: support uniforms" to "grate: support fragment uniforms". We are now use BLOB's symbol location encoding for fragment uniforms, so setting fragment uniform also works for cgc shaders. It is very similar to vertex uniforms, the difference is that BLOB supports only one type of fragment uniform - a vector and it's components enable mask. So a float uniform is a vector with 1 component enabled and mat4 is represented as four vec4's. Please take a look :)

kusma reviewed (approved):
Nice :)
